src/modes/ModeEstimation.jl

```diff
@@ -369,6 +395,11 @@ function _optimize(
     args...; 
     kwargs...
 )
+    # Throw an error if we received a second-order optimizer.
```

Comment:
Do we have to do that? Doesn't Optim just use ForwardDiff (or FD?) to compute the Hessian in this case? If that's the case, then we shouldn't throw an error IMO. It might not be the most efficient approach and would not adhere to the user-provided AD settings but as long as it works we could only print a warning.

Reply:
In FD yes; but (for example in the project that I am working on) it could be that users only define custom adjoints for the gradients but not the Hessian. Therefore even the user provides an AD backend, it might not be a great idea if it by default take that for Hessian function.

Reply:
You mean we shouldn't even print a warning? Would be fine with me as well.

Reply:
Oh, I think throwing an error when some Hessian-required optimizer is received is a great idea, just like what Cameron did here.

Reply:
If we want to throw an error if the Hessian is evaluated, I suggest using `only_fgh!(f)` and implementing `f(F, G, H, x)` that contains the check
```julia
if H !== nothing
    error("second order methods are not supported at the moment")
end
```
In general, this approach is more flexible, avoids baking in a hardcoded check for a special type of a different package in our implementation, and avoids incorrect and unexpected behaviour for second-order optimization algorithms that don't subtype this specific type (since multiple inheritance is not possible in Julia, that's not an impossible scenario per se).

Reply:
I had this, but using `only_fgh!` when the optimizer is first order throws the error

```julia
ERROR: LoadError: MethodError: objects of type Nothing are not callable
Stacktrace:
 [1] (::NLSolversBase.var"#61#62"{NLSolversBase.InplaceObjective{Nothing,Nothing,Turing.OptimLogDensity{DynamicPPL.Model{var"###evaluator#331",(),Tuple{},(),DynamicPPL.ModelGen{var"###generator#332",(),(),Tuple{}}},Turing.OptimizationContext{DynamicPPL.LikelihoodContext{Nothing}},DynamicPPL.VarInfo{NamedTuple{(:s, :m),Tuple{DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:s,Tuple{}},Int64},Array{InverseGamma{Float64},1},Array{DynamicPPL.VarName{:s,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:m,Tuple{}},Int64},Array{Normal{Float64},1},Array{DynamicPPL.VarName{:m,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}}}},Float64}},Nothing,Nothing},Float64})(::Array{Float64,1}, ::Array{Float64,1}) at /home/cameron/.julia/packages/NLSolversBase/mGaJg/src/objective_types/incomplete.jl:45
 [2] value_gradient!!(::OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, ::Array{Float64,1}) at /home/cameron/.julia/packages/NLSolversBase/mGaJg/src/interface.jl:82
 [3] initial_state(::LBFGS{Nothing,LineSearches.InitialStatic{Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.var"#15#17"}
, ::Optim.Options{Float64,Nothing}, ::OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, ::Array{Float64,1}) at /home/cameron/.julia/packages/Optim/TNmSw/src/multivariate/solvers/first_order/l_bfgs.jl:158
 [4] optimize(::OnceDifferentiable{Float64,Array{Float64,1},Array{Float64,1}}, ::Array{Float64,1}, ::LBFGS{Nothing,LineSearches.InitialStatic{
Float64},LineSearches.HagerZhang{Float64,Base.RefValue{Bool}},Optim.var"#15#17"}, ::Optim.Options{Float64,Nothing}) at /home/cameron/.julia/packages/Optim/TNmSw/src/multivariate/optimize/optimize.jl:33
```

@pkofod do you know if this is supposed to happen?

Reply:
At the very least maybe the fix is 
```julia
    M = if optimizer isa Optim.SecondOrderOptimizer
        Optim.optimize(Optim.only_fgh!(f), init_vals, optimizer, options, args...; kwargs...)
    else
        Optim.optimize(Optim.only_fg!(f), init_vals, optimizer, options, args...; kwargs...)
    end
```

Reply:
In addition to throwing the error @devmotion added above, I mean. It seems to be working well on my local run.

Reply:
It's a bug. I think there might be an issue for it.

Reply:
I just found https://github.com/JuliaNLSolvers/Optim.jl/issues/718, I guess that's the related issue.

Reply:
Yeah, I have a fix. Sorry to cossio for waiting a year and a half :grimacing:

Reply:
I mean, I will tag a fix in an hour or so, so please don't special case with a branch.

Reply:
19141